### PR TITLE
MEN-4502: Do not change the underlying Artifact unnecessarily

### PR DESCRIPTION
**Pull request by oleorhagen**
Fixes the issue where modifying an artifact by any of the utility commands would change the compression to `gzip`, as the Artifact was always repacked with the standard compression. This would also destroy signed Artifacts using commands such as `cat` which should have no effect on the signature.

This led to the discovery along the way, that most commands did indeed not respect the `compression` flag - which claimed to be global in scope. Therefore, this removes the global claim from the helptext, and instead adds a note about the `compression` flag's functionality, which after all, if you read the note, was limited to only a few commands in scope.

I guess most people never used the `compression` flag with most of the utility commands (as it would not have worked). Therefore I feel these changes are safe. 

The warning added tells the user to use  the `<modify>` command to change the `compression` type of an image/Artifact. Which I feel is, and should have been, the right approach all along.

Signed-off-by: Ole Petter <ole.orhagen@northern.tech>

### CODE REVIEW

**Comment by oleorhagen** [comment]
@mender-test-bot start pipeline please!

**Review by lluiscampos — COMMENTED** [review]
I am sorry but you need to ELI5 this to me... I don't manage to follow the existing code to see why closing the files is a problem.

**Comment by lluiscampos** [comment]
Thanks for the call, I understood it now.

As discussed: the problem is in `repackArtifact` and we have fixed `cat` and `copy` (out) but just avoiding the close and therefore `repackArtifact` -> However I suspect we have the same problem on `copy` (in), `install`, `delete`, and any other command that has to actually call `repackArtifact`

**Comment by oleorhagen** [comment]
> Thanks for the call, I understood it now.
> 
> As discussed: the problem is in `repackArtifact` and we have fixed `cat` and `copy` (out) but just avoiding the close and therefore `repackArtifact` -> However I suspect we have the same problem on `copy` (in), `install`, `delete`, and any other command that has to actually call `repackArtifact`

Well, not the same problem per se. This is unrelated to the `cat` and `copy` and `modify` commands, and lies deeper in the parser. I've had a look at it, and it is not such an easy fix, since it touches a lot of other parts.

I say we merge this, independently, and then create a follow up for the re-packaging with the proper compression format, in the meantime, it is possible to force the correct compression using something like:
```
mender-artifact cp <from> <to> --compression=<correct-compression>
```
What do you say?

**Comment by oleorhagen** [comment]
Actually on the trail of something now

**Comment by oleorhagen** [comment]
@lluiscampos have a look at the latest 2 commits, and see if you think this is sane or not

**Comment by lluiscampos** [comment]
> @lluiscampos have a look at the latest 2 commits, and see if you think this is sane or not

It makes much sense to me. Getting the compression parameter from the reader is the right thing to do :+1:  I'll have a second read on the code in a short while.

> ...
> What do you say?

Merging as it was would have worked too, you had a valid point.

**Comment by oleorhagen** [comment]
Great, I will probably squash the two into one then, and add a `Changelog` to begin with.

**Comment by oleorhagen** [comment]
Tests will also need a little polishing

File: cli/mender-artifact/copy.go [hunk 1]
```diff
@@ -22,17 +22,12 @@ import (
...
 	"github.com/urfave/cli"
 )
 
 var isimg = regexp.MustCompile(`\.(mender|sdimg|uefiimg|img)`)
 
 func Cat(c *cli.Context) (err error) {
-	comp, err := artifact.NewCompressorFromId(c.GlobalString("compression"))
-	if err != nil {
-		return cli.NewExitError("compressor '"+c.GlobalString("compression")+"' is not supported: "+err.Error(), 1)
-	}
```
Comment (kacf) [inline cli/mender-artifact/copy.go:35]: I think this block (and the other ones you changed) should be replaced with a warning instead then. Something like "`--compression flag ignored, using existing compression type`" or something like that.

File: cli/mender-artifact/artifacts.go [hunk 2]
```diff
@@ -326,7 +326,7 @@ func reconstructArtifactWriteData(ua *unpackedArtifact) (*awriter.WriteArtifactA
 }
 
 func repack(comp artifact.Compressor, ua *unpackedArtifact, to io.Writer, key []byte) error {
-	aWriter := awriter.NewWriter(to, comp)
+	aWriter := awriter.NewWriter(to, ua.ar.Compressor())
```
Comment (kacf) [inline cli/mender-artifact/artifacts.go:329]: Does this prevent repacking with a new compression type using `modify`? Because this is a quite important use case, if you're trying to deploy to a device and find that it's lacking lzma, for example.

File: cli/mender-artifact/copy.go [hunk 3]
```diff
@@ -46,15 +46,6 @@ func Cat(c *cli.Context) (err error) {
...
 	r, err := virtualImage.OpenFile(comp, privateKey, c.Args().First())
-	defer func() {
-		if r == nil {
-			return
-		}
-		cerr := r.Close()
-		if err == nil {
-			err = cerr
-		}
-	}()
```
Comment (kacf) [inline cli/mender-artifact/copy.go:57]: I haven't checked explicitly, but I suspect this will leave all temporary (and very big!) files on the filesystem, so you should check this.

If my suspicion is correct, then a better fix might be to leave this in, but instead have a flag in `r` which indicates whether you want repacking when the handle is closed or not.

File: cli/mender-artifact/copy.go (hunk 1 above)
Comment (oleorhagen) [reply cli/mender-artifact/copy.go:35]: Hmm, yeh, maybe. This one's a little annoying too though, as the flag always has a default `gzip` value, has it not? So the warning will be printed on every single invocation

File: cli/mender-artifact/artifacts.go (hunk 2 above)
Comment (oleorhagen) [reply cli/mender-artifact/artifacts.go:329]: I see, hmm, then we will need a special case for this

File: cli/mender-artifact/copy.go (hunk 3 above)
Comment (oleorhagen) [reply cli/mender-artifact/copy.go:57]: It will -_-
I was hoping for a quick-fix.
Some handlers already have a `dirty` flag set, when closing. I will extend this to all the other handlers too.

File: cli/mender-artifact/copy.go (hunk 1 above)
Comment (kacf) [reply cli/mender-artifact/copy.go:35]: Yeah, but maybe it can be changed to empty default instead? Which is taken into account when the compressor is created.

File: cli/mender-artifact/main.go [hunk 4]
```diff
@@ -478,12 +478,6 @@ func getCliContext() *cli.App {
 		Action: Copy,
 	}
 
-	copy.Flags = []cli.Flag{
-		compressionFlag,
-		privateKeyFlag,
-	}
```
Comment (oleorhagen) [inline cli/mender-artifact/main.go:484]: Hmm, this looks a little too much to me now, looking at it.
I guess at least the `privateKeyFlag` should stay? And possibly the `compressionFlag` is needed too, in order to actually print the warning when it is set?

File: cli/mender-artifact/modify.go [hunk 5]
```diff
@@ -46,7 +46,12 @@ func modifyArtifact(c *cli.Context) (err error) {
 		return cli.NewExitError("File ["+c.Args().First()+"] does not exist.", 1)
 	}
 
-	image, err := virtualImage.Open(comp, privateKey, c.Args().First())
+	var image VPImage
+	if c.String("compression") != "" {
+		image, err = virtualImage.Open(privateKey, c.Args().First(), comp)
+	} else {
+		image, err = virtualImage.Open(privateKey, c.Args().First())
+	}
```
Comment (oleorhagen) [inline cli/mender-artifact/modify.go:54]: Not pretty, but it was the least invasive way I found around this issue

File: cli/mender-artifact/modify.go [hunk 6]
```diff
@@ -62,6 +67,7 @@ func modifyArtifact(c *cli.Context) (err error) {
 		}
 	}()
 
+	image.dirtyImage()
```
Comment (oleorhagen) [inline cli/mender-artifact/modify.go:70]: I decided not to be too surgical with `modify`, as the odds are every command will need the `Artifact` to be repacked.

File: cli/mender-artifact/partition.go [hunk 7]
```diff
@@ -49,6 +49,7 @@ type VPImage interface {
 	io.Closer
 	Open(fpath string) (VPFile, error)
 	OpenDir(fpath string) (VPDir, error)
+	dirtyImage()
```
Comment (oleorhagen) [inline cli/mender-artifact/partition.go:52]: Now only repack images when a command has called `dirtyImage()` as a part of it's execution. This means every command like `CopyTo`, `Write`, etc, all need to call this upon execution.

For this PR, I tried plumbing everything from the image wrappers: `imageAndFile` etc.

File: cli/mender-artifact/partition.go [hunk 8]
```diff
@@ -130,6 +132,13 @@ func (v vImage) Open(comp artifact.Compressor, key []byte, imgname string) (VPIm
 			return nil, errors.Wrap(err, "can not process artifact")
 		}
 
+		var comp artifact.Compressor
+		if len(overrideCompressor) == 1 {
+			comp = overrideCompressor[0]
```
Comment (oleorhagen) [inline cli/mender-artifact/partition.go:137]: Don't really think it's necessary to be too defensive here, we control everywhere this is called.

**Comment by oleorhagen** [comment]
@kacf @lluiscampos I have taken another stab at this (it's been - as always when working with this repo _sigh_ - slightly painful do get anything done without creating a complete mess). 

Please have another look, and tell me what you think of the approach, I've tried to be as clear as I can, but my head's in a little too deep atm, and I'm struggling with getting perspective. Hopefully it will be easier for you guys to see some obvious mistakes.

I've left some comments to help guide you through the review.

Cheers
- Ole uno

File: cli/mender-artifact/main.go (hunk 4 above)
Comment (oleorhagen) [reply cli/mender-artifact/main.go:484]: Well, for now, the `compression` flag does not apply when copying into the Artifact. If compression is set, a warning is issued though, that you should use `mender-artifact modify` to change the compression of your `Artifact`.

File: cli/mender-artifact/modify.go (hunk 5 above)
Comment (oleorhagen) [reply cli/mender-artifact/modify.go:54]: Nope, because that is the compressor is there acquired from the `global` scope variable for `compression`, which defaults to `gzip`.

Comment (kacf) [reply cli/mender-artifact/modify.go:54]: :+1:

File: cli/mender-artifact/main.go (hunk 4 above)
Comment (kacf) [reply cli/mender-artifact/main.go:484]: Ok, sounds good!

Discussed on Slack: Keep both flags, but warn on use of `compression`.

File: cli/mender-artifact/main.go [hunk 9]
```diff
@@ -59,6 +59,38 @@ func main() {
 	}
 }
 
+// Copied from urfave/cli/template.go
```
Comment (lluiscampos) [inline cli/mender-artifact/main.go:62]: Maybe we could have updated [our fork](https://github.com/mendersoftware/cli/commit/2eb686fe967b14b3e204f0eff66427e1fdafcf6d) instead?

Comment (oleorhagen) [reply cli/mender-artifact/main.go:62]: Hmm, possibly, I'm not sure If I think it's a better solution though. Let's try and keep the forks as similar as possible

File: cli/mender-artifact/copy.go [hunk 10]
```diff
@@ -66,9 +62,9 @@ func Cat(c *cli.Context) (err error) {
 }
 
 func Copy(c *cli.Context) (err error) {
-	comp, err := artifact.NewCompressorFromId(c.GlobalString("compression"))
-	if err != nil {
-		return cli.NewExitError("compressor '"+c.GlobalString("compression")+"' is not supported: "+err.Error(), 1)
+	if c.String("compression") != "" {
+		fmt.Fprintf(os.Stderr, "Warning: The compression flag is not respected for the copy command.\n"+
+			"If you whish to change the compression type, use the <modify> command.")
```
Comment (kacf) [inline cli/mender-artifact/copy.go:67]: `s/whish/wish/`

File: cli/mender-artifact/main.go (hunk 9 above)
Comment (kacf) [reply cli/mender-artifact/main.go:62]: Agree, this is a custom template, we should not modify the original I think.

**Review by kacf — APPROVED** [review]
LGTM, just one minor typo.

**Comment by oleorhagen** [comment]
@mender-test-bot start pipeline :robot:

**Comment by oleorhagen** [comment]
Only `test_bogus_proto` test failed in `OS`. 

Good to merge as is.

**Comment by oleorhagen** [comment]
Cherry: 3.5.x: https://github.com/mendersoftware/mender-artifact/pull/334
